File: calculus_and_numerical_methods/newton_raphson.cpp

/**
 * Newton-Raphson Method for Root Finding
 * Author: Scott Douglass
 * License: MIT
 */

#include <iostream>
#include <cmath>

double f(double x) {
    return x*x - 2;
}

double df(double x) {
    return 2*x;
}

double newton(double x0, int max_iter = 20, double tol = 1e-6) {
    double x = x0;
    for (int i = 0; i < max_iter; ++i) {
        double x_new = x - f(x)/df(x);
        if (std::fabs(x_new - x) < tol) return x_new;
        x = x_new;
    }
    return x;
}

int main() {
    std::cout << "Root ≈ " << newton(1.0) << std::endl;
    return 0;
}

File: calculus_and_numerical_methods/newton_raphson.py
"""
newton_raphson.py
Root-finding using the Newton-Raphson method.
Author: Scott
"""

def f(x):
    return x**2 - 2  # Example: root of x^2 - 2 = 0

def df(x):
    return 2 * x  # Derivative

def newton_raphson(x0, tol=1e-6, max_iter=100):
    for i in range(max_iter):
        fx = f(x0)
        dfx = df(x0)
        if dfx == 0:
            raise ZeroDivisionError("Derivative is zero.")
        x1 = x0 - fx / dfx
        if abs(x1 - x0) < tol:
            return x1
        x0 = x1
    raise Exception("Max iterations exceeded")

if __name__ == "__main__":
    x0 = float(input("Initial guess (default 1.0): ") or 1.0)
    root = newton_raphson(x0)
    print(f"Root ≈ {root:.6f}")
